skydb/skydb.go

```diff
@@ -0,0 +1,159 @@
+	}
+	skydEndpoint := os.Getenv("SKYDB_ENDPOINT")
+	if skydEndpoint == "" {
+		return nil, errors.New("missing SKYDB_ENDPOINT environment variable")
+	}
+	opts.Address = skydEndpoint
+	skydb := &SkyDB{Client: &client.Client{opts}}
+	copy(skydb.sk[:], sk)
+	copy(skydb.pk[:], pk)
+	return skydb, nil
+}
+
+// Read retrieves from SkyDB the data that corresponds to the given key set.
+func (db SkyDB) Read(dataKey crypto.Hash) ([]byte, uint64, error) {
+	s, rev, err := registryRead(db.Client, db.pk, dataKey)
+	if err != nil && strings.Contains(err.Error(), "registry entry not found within given time") {
+		return nil, 0, ErrNotFound
+	}
+	if err != nil {
+		return nil, 0, errors.AddContext(err, "failed to read from registry")
+	}
+	b, err := db.Client.SkynetSkylinkGet(s.String())
+	if err != nil {
+		return nil, 0, errors.AddContext(err, "failed to download data from Skynet")
+	}
+	return b, rev, nil
+}
+
+// Write stores the given `data` in SkyDB under the given key set.
+func (db SkyDB) Write(data []byte, dataKey crypto.Hash, rev uint64) error {
+	skylink, err := uploadData(db.Client, data)
+	if err != nil {
+		return errors.AddContext(err, "failed to upload data")
+	}
+	_, err = registryWrite(db.Client, skylink, db.sk, db.pk, dataKey, rev)
+	if err != nil {
+		return errors.AddContext(err, "failed to write to the registry")
+	}
+	return nil
+}
+
+// registryWrite updates the registry entry with the given dataKey to contain the
+// given skylink. Returns a SkylinkV2.
+func registryWrite(c *client.Client, skylink string, sk crypto.SecretKey, pk crypto.PublicKey, dataKey crypto.Hash, rev uint64) (skymodules.Skylink, error) {
+	var sl skymodules.Skylink
+	err := sl.LoadString(skylink)
+	if err != nil {
+		return skymodules.Skylink{}, errors.AddContext(err, "failed to load skylink data")
+	}
+	// Update the registry with that link.
+	spk := types.SiaPublicKey{
+		Algorithm: types.SignatureEd25519,
+		Key:       pk[:],
+	}
+	srv := modules.NewRegistryValue(dataKey, sl.Bytes(), rev).Sign(sk)
+	err = c.RegistryUpdate(spk, dataKey, srv.Revision, srv.Signature, sl)
+	if err != nil {
+		return skymodules.Skylink{}, err
+	}
+	return skymodules.NewSkylinkV2(spk, dataKey), nil
+}
+
+// registryRead reads a registry entry and returns the SkylinkV2 it contains, as well
+// as the revision.
+func registryRead(c *client.Client, pk crypto.PublicKey, dataKey crypto.Hash) (skymodules.Skylink, uint64, error) {
+	spk := types.SiaPublicKey{
+		Algorithm: types.SignatureEd25519,
+		Key:       pk[:],
+	}
+
+	srv, err := c.RegistryRead(spk, dataKey)
+	if err != nil {
+		return skymodules.Skylink{}, 0, errors.AddContext(err, "failed to read from the registry")
+	}
+	err = srv.Verify(pk)
+	if err != nil {
+		return skymodules.Skylink{}, 0, errors.AddContext(err, "the value we read failed validation")
+	}
+	var sl skymodules.Skylink
+	err = sl.LoadBytes(srv.Data)
+	if err != nil {
+		return skymodules.Skylink{}, 0, errors.AddContext(err, "registry value is not a SkylinkV2")
```

Comment:
Shouldn't the registry value be a skylink v1 and not v2? We should also probably add a check.

Reply:
> We care about it here because we are only saving v2 skylinks there.

Aren't we uploading data, getting a v1 skylink and saving the v1 skylink to the registry?

Reply:
I think you are right. We are calling skyd's `SkynetSkyfilePost` method which returns a string skylink, so I would assume a version 1. The `registryWrite` method returns a version 2 skylink but that's hardly relevant because the returned skylink is ignored by the exposed `Write` method.

So, you are right - we are writing a version 1 skylink. I'll change the error message. :)